etl/experiments.kp/knowledge.md

```diff
@@ -0,0 +1,262 @@
+---
+title: Experiment Job
+authors:
+- Frank Bertsch
+tags:
+- experiment
+- firefox
+created_at: 2017-02-01 00:00:00
+updated_at: 2017-02-07 14:16:47.616739
+tldr: 'We take all the pings from yesterday, get the information about any experiments:
+  those that started, those running, and those that ended. These are aggregated by
+  channel and outputted to files in s3.'
+---
+
+```python
+from datetime import datetime as dt, timedelta, date
+import moztelemetry
+from os import environ
+
+# get the desired target date from the environment, or run
+# on 'yesterday' by default.
+yesterday = dt.strftime(dt.utcnow() - timedelta(1), "%Y%m%d")
+target_date = environ.get('date', yesterday)
+```
+
+```python
+from moztelemetry.dataset import Dataset
+
+sample_rate = environ.get('sample', 1)
+pings = Dataset.from_source("telemetry") \
+                   .where(submissionDate=target_date) \
+                   .where(docType="main") \
+                   .where(appName="Firefox") \
+                   .records(sc, sample=sample_rate)
+```
+
+```python
+from moztelemetry import get_pings_properties
+
+subset = get_pings_properties(pings, {
+    "appUpdateChannel": "meta/appUpdateChannel",
+    "log": "payload/log",
+    "activeExperiment": "payload/info/activeExperiment",
+    "activeExperimentBranch": "payload/info/activeExperimentBranch"
+})
+```
+
+```python
+from collections import defaultdict
+
+### Setup data structures and constants ###
+
+ALLOWED_ENTRY_TYPES = ('EXPERIMENT_ACTIVATION', 'EXPERIMENT_TERMINATION')
+
+experiment = {
+    'EXPERIMENT_ACTIVATION': defaultdict(int), 
+    'active': defaultdict(int), 
+    'EXPERIMENT_TERMINATION': defaultdict(int)
+}
+
+channel = {
+    'total': 0, 
+    'errors': [], 
+    'experiments': {}
+}
+
+def get_empty_channel():
+    return deepcopy(channel)
+```
+
+```python
+import gzip
+import ujson
+
+from copy import deepcopy
+
+### Aggregation functions, Spark job, output file creation ###
+
+def channel_ping_agg(channel_agg, ping):
+    """Aggregate a channel with a ping"""
+    channel_agg['total'] += 1
+
+    try:
+        for item in (ping.get("log") or []):
```

Comment:
This isn't the same, right? The ping will always have "log", just sometimes it's an empty dict (resulting from get_pings_properties). Any case where "log" is None, empty, missing, etc. I want it to be [].